### PR TITLE
health: track whether magicsock receive functions are running

### PR DESCRIPTION
**Pull request by josharian**
Review skeptically; I'm not entirely sure I'm holding package health correctly.

### CODE REVIEW

File: health/health.go [hunk 1]
```diff
@@ -263,6 +278,9 @@ func overallErrorLocked() error {
 	_ = lastMapRequestHeard
 
 	var errs []error
+	if !receiveIPv4Running || !receiveDERPRunning {
```
Comment (bradfitz) [inline health/health.go:281]: break this into 3 separate error cases for clarity?

and then you'd need to know from magicsock whether IPv6 was wanted.

**Review by bradfitz — APPROVED** [review]

**Comment by josharian** [comment]
PTAL

File: health/health.go [hunk 2]
```diff
@@ -263,6 +280,15 @@ func overallErrorLocked() error {
 	_ = lastMapRequestHeard
 
 	var errs []error
+	if !receiveIPv4Running {
+		errs = append(errs, errors.New("receiveIPv4 not running"))
+	}
+	if !receiveDERPRunning {
+		errs = append(errs, errors.New("receiveDERP not running"))
+	}
+	if receiveIPv6Started || !receiveIPv6Running {
```
Comment (bradfitz) [inline health/health.go:289]: I think you meant && here and not ||

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -1581,6 +1581,9 @@ func (c *Conn) noteRecvActivityFromEndpoint(e conn.Endpoint) {
 
 // receiveIPv6 receives a UDP IPv6 packet. It is called by wireguard-go.
 func (c *Conn) receiveIPv6(b []byte) (int, conn.Endpoint, error) {
+	health.SetReceiveIPv6Started(true)
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:1584]: I expected you to call this in connBind.Open, at:

```go
        if c.pconn6 != nil {
                fns = append(fns, c.receiveIPv6)
        }
```

Comment (josharian) [reply wgengine/magicsock/magicsock.go:1584]: Then there's a (bigger) window during which you can get a false positive. Plus for anyone reading the code later, it makes it totally obvious that `started && !running` occurs when this function exits.

File: health/health.go (hunk 2 above)
Comment (josharian) [reply health/health.go]: Oops, thanks.

File: wgengine/magicsock/magicsock.go (hunk 3 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:1584]: Ah, okay, now I see.

**Review by bradfitz — APPROVED** [review]

**Pull request merged**